Source/Dafny/Compiler-Csharp.cs

```diff
@@ -1709,11 +1721,41 @@ protected class ClassWriter : IClassWriter {
     }
 
     protected override void EmitSeqConstructionExpr(SeqConstructionExpr expr, bool inLetExprBody, TargetWriter wr) {
-      wr.Write("{0}<{1}>.Create(", DafnySeqClass, TypeName(expr.Type.AsSeqType.Arg, wr, expr.tok));
-      TrExpr(expr.N, wr, inLetExprBody);
-      wr.Write(", ");
-      TrExpr(expr.Initializer, wr, inLetExprBody);
-      wr.Write(")");
+      var elmtType = expr.Type.AsSeqType.Arg;
+      if (expr.Initializer is LambdaExpr lam) {
+        Contract.Assert(lam.BoundVars.Count == 1);
+
+        wr.Write("((System.Func<{0}>) (() =>", TypeName(expr.Type, wr, expr.tok));
+        var wrLamBody = wr.NewBigExprBlock("", "))()");
+
+        var indexType = expr.N.Type;
+        var lengthVar = FreshId("dim");
+        DeclareLocalVar(lengthVar, indexType, expr.tok, expr.N, inLetExprBody, wrLamBody);
+        var arrVar = FreshId("arr");
+        wrLamBody.Write("var {0} = ", arrVar);
+        var wrDims = EmitNewArray(elmtType, expr.tok, dimCount: 1, mustInitialize: false, wr: wrLamBody);
+        Contract.Assert(wrDims.Count == 1);
+        wrDims[0].Write(lengthVar);
+        wrLamBody.WriteLine(";");
+
+        var intIxVar = FreshId("i");
+        var wrLoopBody = wrLamBody.NewBlockWithPrefix("for (int {0} = 0; {0} < {1}; {0}++)", "", intIxVar, lengthVar);
```

Comment:
`lengthVar` is of type `indexType`, which is probably `BigInteger`. If so, the comparison `{0} < {1}` will compare an `int` with ` BigInteger`. This is okay, but could be slightly improved by first sticking `lengthVar` into a `uint`.

Reply:
For that matter, should `i` be a `long`?  There are other ways we support (in C#) sequences of 64-bit lengths.